Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp

```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+            return (static_cast<uint64_t>(hi) << 32) | lo;
+        };
+
+        auto parseTimeValue = [&](uint32_t sec_lo, uint32_t sec_hi, uint32_t nsec) -> Seconds {
+            return Seconds { static_cast<double>(combineToUInt64(sec_lo, sec_hi)) } + Seconds::fromNanoseconds(nsec);
+        };
+
+        // Quoting https://wayland.app/protocols/presentation-time#wp_presentation_feedback:event:presented:
+        // The associated content update was displayed to the user at the indicated time (tv_sec_hi/lo, tv_nsec).
+        // For the interpretation of the timestamp, see presentation.clock_id event.
+        auto presentationTime = parseTimeValue(tv_sec_lo, tv_sec_hi, tv_nsec);
+        auto refreshInterval = Seconds::fromNanoseconds(refresh_nsec);
+        auto sequence = combineToUInt64(seq_lo, seq_hi);
+        priv->presentationFeedbackStatistics->presentedFrame(feedback, presentationTime, refreshInterval, sequence, flags);
+    },
+    // discarded
+    [](void* data, struct wp_presentation_feedback* feedback) {
+        auto* view = WPE_VIEW(data);
+        auto* priv = WPE_VIEW_WAYLAND(view)->priv;
+
+        // Quoting https://wayland.app/protocols/presentation-time#wp_presentation_feedback:event:discarded:
+        // The content update was never displayed to the user.
+        priv->presentationFeedbackStatistics->discardedFrame(feedback);
+    },
+};
+// END
+
+PresentationFeedbackStatistics::PresentationFeedbackStatistics(unsigned historySize)
+    : m_maxHistorySize(historySize)
+{
+}
+
+PresentationFeedbackStatistics::~PresentationFeedbackStatistics()
+{
+    for (auto [presentationFeedback, frameStartTime] : m_pendingFeedbacks)
+        wp_presentation_feedback_destroy(presentationFeedback);
+}
+
+void PresentationFeedbackStatistics::beginFrame(struct wp_presentation* presentation, WPEView* view)
+{
+    auto* wlSurface = wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(view));
+
+    auto* presentationFeedback = wp_presentation_feedback(presentation, wlSurface);
+    wp_presentation_feedback_add_listener(presentationFeedback, &presentationFeedbackListener, view);
+    m_pendingFeedbacks.append({ presentationFeedback, MonotonicTime::now() });
+
+    WTFEmitSignpost(this, WaylandFrameBegin, "pending feedbacks? %lu", m_pendingFeedbacks.size());
+}
+
+void PresentationFeedbackStatistics::presentedFrame(struct wp_presentation_feedback* feedback, Seconds presentationTime, Seconds refreshInterval, uint64_t sequence, uint32_t flags)
+{
+    std::optional<Seconds> frameStartTime;
+    for (size_t i = 0; i < m_pendingFeedbacks.size(); ++i) {
+        auto [presentationFeedback, frameStartTimeForFeedback] = m_pendingFeedbacks[i];
+        if (presentationFeedback == feedback) {
+            frameStartTime = frameStartTimeForFeedback.secondsSinceEpoch();
+            wp_presentation_feedback_destroy(feedback);
+            m_pendingFeedbacks.remove(i);
+            break;
+        }
+    }
```

Comment:
I think you can use removeFirstMatching() capturing frameStartTime in the lambda and calling wp_presentation_feedback_destroy() from the lambda too.